kernel/aster-nix/src/process/posix_thread/futex.rs

```diff
@@ -257,72 +268,83 @@ impl FutexBucket {
         max_nrequeues: usize,
     ) {
         let mut count = 0;
+        let mut item_cursor = self.items.front_mut();
+        while !item_cursor.is_null() {
+            if count >= max_nrequeues {
+                break;
+            }
 
-        self.queue.retain(|item| {
-            if count >= max_nrequeues || key != item.key {
-                true
-            } else {
-                let mut new_item = item.clone();
-                new_item.key = new_key;
-                another.enqueue_item(new_item);
+            if item_cursor.get().is_some_and(|item| item.key == key) {
+                let mut item = item_cursor.remove().unwrap();
+                item.key = new_key;
+                another.add_item(item);
                 count += 1;
-                false
+            } else {
+                item_cursor.move_next()
             }
-        });
+        }
     }
 }
 
-#[derive(Debug, PartialEq, Clone)]
 struct FutexItem {
     key: FutexKey,
-    bitset: FutexBitSet,
-    waiter: FutexWaiterRef,
+    waker: Arc<Waker>,
+    link: LinkedListAtomicLink,
+}
+
+impl PartialEq for FutexItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key
+    }
 }
 
 impl FutexItem {
-    pub fn new(key: FutexKey, bitset: FutexBitSet) -> Self {
-        FutexItem {
+    pub fn build(key: FutexKey) -> (Waiter, Box<Self>) {
+        let (waiter, waker) = Waiter::new_pair();
+        let futex_item = Box::new(FutexItem {
             key,
-            bitset,
-            waiter: Arc::new(FutexWaiter::new()),
-        }
-    }
+            waker,
+            link: LinkedListAtomicLink::new(),
+        });
 
-    pub fn wake(&self) {
-        self.waiter.wake();
+        (waiter, futex_item)
     }
 
-    pub fn wait(&self, timeout: Option<FutexTimeout>) {
-        self.waiter.wait(timeout);
+    pub fn wake(&self) {
+        self.waker.wake_up();
     }
+}
 
-    pub fn waiter(&self) -> &FutexWaiterRef {
-        &self.waiter
-    }
+// The addr of a futex, it should be used to mark different futex word
+#[derive(Debug, Clone, Copy)]
+struct FutexKey {
+    addr: Vaddr,
+    bitset: FutexBitSet,
+}
 
-    pub fn batch_wake(items: &[FutexItem]) {
-        let waiters = items.iter().map(|item| item.waiter()).collect::<Vec<_>>();
-        FutexWaiter::batch_wake(&waiters);
+impl PartialEq for FutexKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.addr == other.addr && (self.bitset & other.bitset) != 0
```

Comment:
Do not implement `PartialEq` this way. Add a separate `match` method instead.

```rust
impl FutexKey {
    fn match(&self, another: &Self) -> bool {
        self.addr == other.addr && (self.bitset & other.bitset) != 0
    }
}
```

Here is the reason against the idea of implementing `PartialEq` this way. Semantically, two sets having an non-empty intersection is not what we call equality. Two sets having exactly the same elements are know as equality. So it is quite surprising to see such an implementation.

More importantly, this implementation is logically flawed. Consider the following three `FutexKey`.

```rust
let addr = 0xdeadbeaf;
let key0 = FutexKey {
    addr,
    bitset: 0b01,
};
let key1 = FutexKey {
    addr,
    bitset: 0b11,
};
let key2 = FutexKey {
    addr,
    bitset: 0b10,
};
```

Now we have the following surprising results when testing equallity.

```rust
assert!(key0 == key1); // Good
assert!(key1 == key2); // No problem
assert!(key2 == key0); // PANIC!
```

But `PartialEq` should ensure _transitivity_, that is if `a == b` and `b == c`, then `a == c`.